packages/kbn-grid-layout/grid/use_grid_layout_events/use_grid_layout_events.ts

```diff
@@ -0,0 +1,285 @@
+        maxColumn
+      );
+      const targetRow = Math.max(Math.round(localYCoordinate / (rowHeight + gutterSize)), 0);
+
+      const requestedGridData = { ...currentGridData };
+      if (isResize) {
+        requestedGridData.width = Math.max(targetColumn - requestedGridData.column, 1);
+        requestedGridData.height = Math.max(targetRow - requestedGridData.row, 1);
+      } else {
+        requestedGridData.column = targetColumn;
+        requestedGridData.row = targetRow;
+      }
+
+      // resolve the new grid layout
+      if (
+        hasChangedGridRow ||
+        !isGridDataEqual(requestedGridData, lastRequestedPanelPosition.current)
+      ) {
+        lastRequestedPanelPosition.current = { ...requestedGridData };
+
+        // remove the panel from the row it's currently in.
+        const nextLayout = currentLayout.map((row) => {
+          const { [interactionEvent.id]: interactingPanel, ...otherPanels } = row.panels;
+          return { ...row, panels: { ...otherPanels } };
+        });
+
+        // resolve destination grid
+        const destinationGrid = nextLayout[targetRowIndex];
+        const resolvedDestinationGrid = resolveGridRow(destinationGrid, requestedGridData);
+        nextLayout[targetRowIndex] = resolvedDestinationGrid;
+
+        // resolve origin grid
+        if (hasChangedGridRow) {
+          const originGrid = nextLayout[lastRowIndex];
+          const resolvedOriginGrid = resolveGridRow(originGrid);
+          nextLayout[lastRowIndex] = resolvedOriginGrid;
+        }
+        if (!deepEqual(currentLayout, nextLayout)) {
+          gridLayout$.next(nextLayout);
+        }
+      }
+    },
+    [gridLayoutStateManager, scrollInterval]
+  );
+
+  const attachMouseEvents = useCallback(
+    (e: UserMouseEvent) => {
+      if (e.button !== MOUSE_BUTTON_LEFT) return;
+
+      const onDragEnd = () => {
+        document.removeEventListener('scroll', onPointerMove);
+        document.removeEventListener('mousemove', onPointerMove);
+
+        stopAutoScroll(scrollInterval);
+        drop(gridLayoutStateManager);
+        onInteractionEvent('drop');
+      };
+
+      document.addEventListener('scroll', onPointerMove);
+      document.addEventListener('mousemove', onPointerMove);
+      document.addEventListener('mouseup', onDragEnd, { once: true });
+    },
+    [onPointerMove, onInteractionEvent, gridLayoutStateManager]
+  );
+
+  const attachTouchEvents = useCallback(
+    (e: UserTouchEvent) => {
+      if (e.touches.length > 1) return;
+
+      const onDragEnd = () => {
+        e.target!.removeEventListener('touchmove', onPointerMove);
+
+        drop(gridLayoutStateManager);
+        onInteractionEvent('drop');
+      };
+
+      e.target!.addEventListener('touchmove', onPointerMove, { passive: false });
+      e.target!.addEventListener('touchend', onDragEnd, { once: true });
+    },
+    [gridLayoutStateManager, onPointerMove, onInteractionEvent]
+  );
+
+  const attachLayoutEvents = useCallback(
```

Comment:
this is the core change - this is the event that we're attaching on `dragStart` ( I am unsure about the current name). It checks if the layout is interactive, then attaches all the other events (moving, scrolling and dropping interaction) and interacts with the `gridLayoutStateManager`. With this state it's gonna be easy to add a keyboard sensor too.